libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp

```diff
@@ -0,0 +1,142 @@
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+  backend driver for airspeed from www.qio-tek.com a I2C ASP5033 sensor
+ */
+
+#include "AP_Airspeed_ASP5033.h"
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Common/AP_Common.h>
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+#include <AP_Math/AP_Math.h>
+
+extern const AP_HAL::HAL &hal;
+
+#define ASP5033_I2C_ADDR 0x6D
+#define SYS_CONFIG       0xA5
+#define CMD              0x30
+#define CMD_MEASURE      0x0A
+#define PRESS_DATA       0x06
+#define TEMP_DATA        0x09
+#define WAI_REG          0x01
+#define WAI_SET          0XA4
+#define DEVICE_ID        0x66
+#define SENSOR_READY     0x08
+
+
+AP_Airspeed_ASP5033::AP_Airspeed_ASP5033(AP_Airspeed &_frontend, uint8_t _instance) :
+    AP_Airspeed_Backend(_frontend, _instance)
+{
+}
+
+// probe and initialise the sensor
+bool AP_Airspeed_ASP5033::init()
+{
+    uint8_t SysID;
+    uint8_t whoami;
+    _dev = hal.i2c_mgr->get_device(get_bus(), ASP5033_I2C_ADDR);
+    if (!_dev) {
+        return false;
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO,"ASP5033: sensor not found\n");
+    }
+    _dev->get_semaphore()->take_blocking();
+    _dev->set_speed(AP_HAL::Device::SPEED_HIGH);
+    _dev->write_register(WAI_SET, DEVICE_ID); //Set Whoami ID. 0XA4 is a OPT WhoamI Part ID register, also can be read from address 0x01.
+    //check whoami, set the PartID register to 0X66 first, them Read the PartID register.
+    if (!_dev->read_registers(WAI_REG, &whoami, 1) || //Read the WhoamI ID
+        whoami != DEVICE_ID){
+        _dev->get_semaphore()->give();
+        return false;
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO,"ASP5033: sensor not found\n"); 
+    }
+    _dev->set_retries(2);
+    _dev->get_semaphore()->give();
+    _dev->read_registers(SYS_CONFIG, &SysID, 1); //Read Sensor_config
+    SysID = (SysID&0xFD); //Read the 0xA5 register value, put the read binary value "and" on "11111111101" then write to 0xA5.
+    _dev->write_register(SYS_CONFIG, SysID); //Set Sensor_config
+    _dev->register_periodic_callback(1000000UL/80U,
+                                    FUNCTOR_BIND_MEMBER(&AP_Airspeed_ASP5033::timer, void));
+    return true;
+}         
+
+// read the data from the sensor
+void AP_Airspeed_ASP5033::timer()
+{
+    uint8_t SoC; //Start of conversion flag
+    uint8_t pressraw_bytes[3]; //pressure data AD value
+    uint8_t tempraw_bytes[2]; //temperature data AD value
+    _dev->write_register(CMD, CMD_MEASURE); //once temperature conversion immediately followed by once sensor signal conversion
+    _dev->read_registers(CMD, &SoC, 1); //Judge whether Data collection is over
```

Comment:
You have to check if the read worked, otherwise SoC is just filled with garbage from the stack, and it may or may not look ready.

Reply:
All registers read checking will be add, Thank you for your advice.